gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopyManifest.java

```diff
@@ -0,0 +1,150 @@
+import com.google.gson.stream.JsonReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+
+/**
+ * Copy Manifest schema and serDe for manifest based copy
+ * https://iwww.corp.linkedin.com/wiki/cf/display/ENGS/Manifest+based+distcp+runbook
+ */
+public class CopyManifest {
+  private static final String MISSING_FN_MESSAGE = "fileName cannot be null";
+  private static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();
+  private static final Type CopyableUnitListType = new TypeToken<ArrayList<CopyableUnit>>(){}.getType();
+
+  public final ArrayList<CopyableUnit> _copyableUnits;
+
+  public CopyManifest() {
+    _copyableUnits = new ArrayList<>();
+  }
+
+  public CopyManifest(ArrayList<CopyableUnit> copyableUnits) {
+    _copyableUnits = copyableUnits;
+  }
+
+  /**
+   * Add a new copyable unit to a copy manifest. Used for building a manifest
+   * @param copyableUnit
+   */
+  public void add(CopyManifest.CopyableUnit copyableUnit) {
+    _copyableUnits.add(copyableUnit);
+  }
+
+  /**
+   * One item in a copy manifest
+   * Only filename is required
+   */
+  public static class CopyableUnit {
+    public final String fileName;
+    public final String fileGroup;
+    public final Long fileSizeInBytes;
+    public final Long fileModificationTime;
+
+    public CopyableUnit(String fileName, String fileGroup, Long fileSizeInBytes, Long fileModificationTime) {
+      this.fileName = fileName;
+      this.fileGroup = fileGroup;
+      this.fileSizeInBytes = fileSizeInBytes;
+      this.fileModificationTime = fileModificationTime;
+      if (this.fileName == null) {
+        throw new IllegalArgumentException(MISSING_FN_MESSAGE);
+      }
+    }
+  }
+
+  /**
+   * Note: naive read does not do validation of schema. For schema validation use CopyableUnitIterator
+   * @param fs filsystem object used for accessing the filesystem
+   * @param path path manifest file location
+   * @return a copy manifest object from the json representation at path
+   * @throws IOException
+   */
+  public static CopyManifest read(FileSystem fs, Path path) throws IOException {
+    JsonReader jsonReader = new JsonReader(new InputStreamReader(fs.open(path), "UTF-8"));
+    return new CopyManifest(GSON.fromJson(jsonReader, CopyableUnitListType));
+  }
+
+  /**
+   *
+   * @param fs filsystem object used for accessing the filesystem
+   * @param path path manifest file location
+   * @throws IOException
+   */
+  public void write(FileSystem fs, Path path) throws IOException {
+    String outputJson = GSON.toJson(this._copyableUnits, CopyableUnitListType);
+    FSDataOutputStream out =  fs.create(path, true);
+    out.write(outputJson.getBytes(StandardCharsets.UTF_8));
```

Comment:
Should we surround the write method with try clause and close the output stream in final block?

Reply:
done